File: 2019/src/main/java/frc/robot/commands/LightColour.java
package frc.robot.commands;

import edu.wpi.first.wpilibj.command.Command;
import frc.robot.Robot;

public class LightColour extends Command {
	private double m_r, m_g, m_b;

	public LightColour(double r, double g, double b) {
		requires(Robot.m_lights);
		m_r = r;
		m_g = g;
		m_b = b;
	}
	
	public void setColour(String hex) {
		requires(Robot.m_lights);
		m_r = (double) Integer.valueOf(hex.substring(1, 3), 16) / 255.0;
		m_b = (double) Integer.valueOf(hex.substring(3, 5), 16) / 255.0;
		m_g = (double) Integer.valueOf(hex.substring(5, 7), 16) / 255.0;
	}

	@Override
	protected void initialize() {
	}

	@Override
	protected void execute() {
		Robot.m_lights.setColour(m_r, m_g, m_b);
	}

	@Override
	protected boolean isFinished() {
		return false;
	}

	@Override
	protected void end() {
		Robot.m_lights.setColour(0.0, 0.0, 0.0);
	}

	@Override
	protected void interrupted() {
		end();
	}
}

File: 2019_cpp/src/main/cpp/commands/LightColour.cpp
#include "commands/LightColour.h"

#include "Robot.h"

LightColour::LightColour(const double &r, const double &g, const double &b) :
	m_r(0.0),
	m_g(0.0),
	m_b(0.0)
{
	Requires(&Robot::m_lights);
}

LightColour::LightColour(const std::string &hex) :
	m_r(0.0),
	m_g(0.0),
	m_b(0.0)
{
	Requires(&Robot::m_lights);

	uint32_t r, g, b;
	sscanf(hex.c_str() + (hex[0] == '#' ? 1 : 0), "%2x%2x%2x", &r, &g, &b);
	m_r = static_cast<double>(r) / 255.0;
	m_g = static_cast<double>(g) / 255.0;
	m_b = static_cast<double>(b) / 255.0;
}

void LightColour::Initialize()
{
}

void LightColour::Execute()
{
	Robot::m_lights.SetColour(m_r, m_g, m_b);
}

bool LightColour::IsFinished()
{ 
	return false; 
}

void LightColour::End()
{
	Robot::m_lights.SetColour(0.0, 0.0, 0.0);
}

void LightColour::Interrupted()
{
	End();
}
